### PR TITLE
Callback restructure and a posteriori test

### LINKED ISSUE #55: Callback history reset
When training using a callback function, it keeps the same history forever. We should add a flag/method to clean the history

### LINKED ISSUE #66: Check efficacy of a posteriori training
We have to confirm that the a posteriori training is effectively going in the right direction. The loss does not decrease enough to confirm this and it is very unstable suggesting some problems.
A basic test could be
- [ ] confirm that a posteriori training converges to something better than no closure

then later we have to confirm that
- [ ] given a specific CNN architecture and data structure, CoupleNODE produces results comparable to NeuralClosure

### PR DESCRIPTION
**Pull request by SCiarella**
This PR closes #55 and address #66 by implementing a new callback function.

The callback function is now created from the constructor `create_callback` that resets the history every time the constructor is called, thus closing #55.
Then the callback function itself can return the history if the argument `return_lhist` is set to `true`. The history can be passed again to the constructor if one wants to change the callback but keep the history.

To address #66, the new callback reports also a 'validation_loss' that is measured over a fixed piece of trajectory and shows the overall effect of the training without being victim of the random selection of a sample at each step. This loss is also compared to the `no_closure` model (LES).
Unfortunately, this shows that our example of a posteriori training is **not** good enough and we can not confirm that our implementation of a posteriori fitting is bug free.
So this PR will not close #66 but it implements some life improvements to continue the work.

-----
Additionally the dataloader and the generate data scripts have been modified in order to allow the use of more than 1 LES trajectories. This change looks fundamental to me, but it may break some symmetry with NeuralClosure and something else that we would like to do.

Finally notice that the a priori fitting was not touched by this PR and it may require some update

### CODE REVIEW

**Comment by luisaforozco** [comment]
This looks good to me but haven't tested it. Have you tested if these changes may break the training of a-priori?
I this is still working and the a-posteriori fitting too, then I am ok with merging it. Tomorrow I can test the code locally.

**Comment by SCiarella** [comment]
> This looks good to me but haven't tested it. Have you tested if these changes may break the training of a-priori? I this is still working and the a-posteriori fitting too, then I am ok with merging it. Tomorrow I can test the code locally.

After the last commit, both the a-priori and a-posteriori training are working

File: simulations/NavierStokes_2D/scripts/preprocess_posteriori.jl [hunk 1]
```diff
@@ -21,6 +21,7 @@ using CoupledNODE.NavierStokes: create_io_arrays_posteriori
 io_post = create_io_arrays_posteriori(data, setups)
 
 # Example of dimensions and how to operate with io_arrays_posteriori
+ig = 1
```
Comment (luisaforozco) [inline simulations/NavierStokes_2D/scripts/preprocess_posteriori.jl:24]: I would avoid defining this here. 
The way in which I thought one would interact with this file is including it from `train_posteriori.jl`, where this variable is define. If we redefine it here, there will be a clash.

File: simulations/NavierStokes_2D/scripts/preprocess_posteriori.jl [hunk 2]
```diff
@@ -30,4 +31,8 @@ io_post[ig].t[2, :]
 # * Create dataloader containing trajectories with the specified nunroll
 using CoupledNODE.NavierStokes: create_dataloader_posteriori
 nunroll = 5
-dataloader_posteriori = create_dataloader_posteriori(io_post[ig]; nunroll = nunroll, rng)
+dataloader_posteriori = create_dataloader_posteriori(io_post; nunroll = nunroll, rng)
+
+# Load the test data
+test_data = load("simulations/NavierStokes_2D/data/data_test.jld2", "data_test")
+test_io_post = create_io_arrays_posteriori(test_data, setups)
```
Comment (luisaforozco) [inline simulations/NavierStokes_2D/scripts/preprocess_posteriori.jl:38]: This leads to a problem. Indeed `create_io_arrays_posteriori ` considers an array with multiple `NamedTuple` in it, each one representing a sample. 
For the *train* dataset, we have multiple samples, but for the *test* dataset we only have one sample and so `test_data` is directly a `NamedTuple` and not an `Array`.

I think we should not try to have special if statements to check the nature of the data here, but in `generate_data.jl` make sure that the format is `Array` even when there is a single sample. What do you think?

Comment (luisaforozco) [reply simulations/NavierStokes_2D/scripts/preprocess_posteriori.jl:38]: Outdated comment. In `generate_data.jl` the *test* data is also an `Array`.

File: simulations/NavierStokes_2D/scripts/train_posteriori.jl [hunk 3]
```diff
@@ -13,15 +13,41 @@ rng = Random.Xoshiro(123)
...
 )
+layers = (
+    Lux.SkipConnection(AttentionLayer(N, d, emb_size, patch_size, n_heads; T = T),
+        (x, y) -> cat(x, y; dims = 3); name = "Attention"),
+    CnnLayers
+)
+closure = Lux.Chain(layers...)
+θ, st = Lux.setup(rng, closure)
+using ComponentArrays: ComponentArray
+θ = ComponentArray(θ)
```
Comment (luisaforozco) [inline simulations/NavierStokes_2D/scripts/train_posteriori.jl:47]: Why do we need this additional block of code?
it seems to me that we should bstract this in the creation of `cnn` or propose another model `cnn_with_attention`. In any case, the same code for generating the model has to be run in the `evaluate_posteriori.jl` so that when loading the model we can apply the weights, avoiding incompatibilities.
